Port tests to JUnit for java.util

ekrich opened the pull request.

ekrich commented:
@sjrd Some of these tests came from Scala.js with names like the following:
```scala
@Test def should_iterate_in__after_accessing_elements()
```
That one has two underscores.

They were changed to this for Scala Native.
```scala
test("should iterate in after accessing elements")
```
Now I made them toggle cased by default but then I noticed how they were in Scala.js. 
```Scala
@Test def shouldIterateInAfterAccessingElements()
```

Should I change them back so they match?

I could also leave them toggle case and then just go back and change them in Scala.js to match - when we added `IdentityHashMap` we made them toggle cased and I think for the most part that is the standard. I don't mind to do an extra PR to make them toggle cased in Scala.js

sjrd commented:
The name formats clearly lack uniformity. New tests should be written `methodName` or `testMethodName` for a test that tests the method `methodName`. But it's not worth going back and changing the name of existing tests. Unless you want to do it over the entire Scala.js test suite to make it 100% uniform everywhere.

ekrich commented:
Ok, for porting then we should make them toggle case. I can look at the scope of the ones in Scala.js - I'm not sure if it is isolated or all over the place.

LeeTibbert commented:
@ekrich I did several spot checks, not line by line but enough to give me the flavor.

I like what you are doing and your attention to detail. The changes make sense to me and
are, IMHO, in the right direction.  Thank you for taking on this effort.

ekrich commented:
Nope got a few more to go.

LeeTibbert commented on unit-tests/src/test/scala/java/util/function/BiConsumerTest.scala:
```diff
@@ -25,14 +30,12 @@ class BiConsumerTest {
-    // result = 0
-    // assertTrue(result == 0)
-    // addT.andThen(timesT).accept(2, 3)
+    assertTrue(result == 0)
```
Given the `@Before`, (clever that!), this seems excessive & redundant.  Yes, some day somebody could remove the `@Before` but then they will get hosed/allerted  by the 
test failing.

LeeTibbert commented on unit-tests/src/test/scala/java/util/function/BiConsumerTest.scala:
```diff
@@ -25,14 +30,12 @@ class BiConsumerTest {
-    // // addT will add 2 + 2 = 4 to result which is zero = 4
-    // // then 2*3 = 6 will be added to result which is 4, =  11
-    // assertTrue(result == 11)
+    // addT will add 2 + 2 = 4 to result which is zero = 4
```
I suggest deleting the comment lines.  What was it that Neils Bohr said about a theory that was so off that it was not even wrong?    The arguments given in this case are 2 and 3, not 2 and 2.  So the sum is 5, added to the (2 * 3 === 6) gives the expected 11.
By my reading, the assert is right and the comment is total misdirection; a polite way of saying so wrong as to waste people's time.

LeeTibbert replied:
Thank you for studying, fixing, & enabling the AndThen test.  It adds assurance that the javalib code is working.

ekrich replied:
Ok, I can fix that but I can't fix everything.

LeeTibbert replied:
Understood. The value of the whole is much better than the individual parts.  
Thank you for removing that comment.  It was so far off that it surely would have
snagged somebody or somebodies, probably us in 6 month time.

ekrich replied:
Sure, no problem. It was pretty much wrong - I was just happy the tests passed. I have tried to clean things up a bit but have tried to just do a direct port as to not confuse matters.

ekrich commented:
I ported and tested each test individually to retain some sense of sanity.

I admit this maybe hard to review based on the sheer bulk of changes. Most of the other packages are smaller.

LeeTibbert commented:
This is a huge PR but I concur that it should go in as soon as feasible.  
It will quickly become overcome by events otherwise.

Travis CI has not completed, so I do not have that evidence. 

So far I have examined somewhere around 10 files.  The changes in each make 
sense.  I especially like the way that the `@ignore` statements were improved.

I understand and appreciate that this was a one-for-one straight port, with content
edits being kept to a minimum.  Perhaps someday either Eric or I could teach 
an extension course in testing and use these tests as a real life basis for 
improvement.  Plenty of work for the next evolution, but IMO, it should no slow
down or delay this evolution.   

I am happy with this PR to date.  I will continue examining files but I do not
want to delay this PR being merged, once Travis is all Green.

Eric, from one community member to another, job well done.  

I am using this work, and some of the JUnit idioms, as a model for my
work writing and porting other Scala Native tests to JUnit.

sjrd commented:
Thank you for this huge undertaking.

This seems virtually impossible to review line by line, especially the files where git doesn't figure out that they're the same. What should I pay attention to?

Also, it seems that the main reason why git can't figure out the renames is because all the `assert`s became `assertTrue`. This would make sense if they were all *supposed* to be `assertTrue`, but many (most) of them should be `assertEquals` in an ideal world. Since `assertTrue` isn't ideal anyway, I suggest keeping the `assert`s in order to help git.

ekrich commented:
@sjrd I did an audit of the log files and found 50 tests searching for `java.util`. I was able to go through the tests in the file system and find 50 tests as well to check them off.

There are a couple of helper classes and traits/abstract classes for collections. `LinkedHashMap...` runs four different variants. I also converted `EmptyPackageTest` which makes a 51st test.

I found a couple of small things that I fixed in the last 2 commits.

ekrich commented:
@sjrd After fixing the 2.11 failures, I have really high confidence in the PR. Maybe poking around is okay to make sure nothing obvious is weird.

My suggestion for the future is that we port the tests without changing the file name but change the internal test name and then once we are done reviewing, change the file names. This way we should be able to just see the diffs and then the rename is pretty easy to see.

sjrd reviewed (approved):
OK, let's take this in before it bit-rots.

ekrich commented:
@sjrd Thank you very much. I’ll jump on the rest. Is the unified diff ok or do you like the two step process with the rename after the first review? Unfortunately git doesn’t handle it super well.

sjrd commented:
Either one is fine. Use the one that is easier for you.
